File: explore/plots/spatiality.R
#!/usr/bin/env Rscript

# DXT Explorer Copyright (c) 2021, The Regents of the University of
# California, through Lawrence Berkeley National Laboratory (subject
# to receipt of any required approvals from the U.S. Dept. of Energy). 
# All rights reserved.
# 
# If you have questions about your rights to use or distribute this software,
# please contact Berkeley Lab's Intellectual Property Office at
# IPO@lbl.gov.
# 
# NOTICE.  This Software was developed under funding from the U.S. Department
# of Energy and the U.S. Government consequently retains certain rights.  As
# such, the U.S. Government has been granted for itself and others acting on
# its behalf a paid-up, nonexclusive, irrevocable, worldwide license in the
# Software to reproduce, distribute copies to the public, prepare derivative 
# works, and perform publicly and display publicly, and to permit others to do so.

packages <- c(
    'png',
    'ggplot2',
    'optparse',
    'plyr',
    'plotly',
    'rmarkdown',
    'htmlwidgets',
    'wesanderson',
    'R.utils',
    'Cairo'
)

# Install packages not yet installed
installed_packages <- packages %in% rownames(installed.packages())

dir.create(path = Sys.getenv("R_LIBS_USER"), showWarnings = FALSE, recursive = TRUE)

if (any(installed_packages == FALSE)) {
    install.packages(packages[!installed_packages], repos='http://cran.us.r-project.org', quiet=TRUE, lib=Sys.getenv("R_LIBS_USER"))
}

# Packages loading
invisible(lapply(packages, library, warn.conflicts = FALSE, quietly = TRUE, character.only = TRUE))

if (pandoc_available()) {
    self_contained = TRUE
} else {
    self_contained = FALSE
}

option_list = list(
    make_option(
        c('-f', '--file'),
        type = 'character',
        default = NULL, 
        help = 'DXT CSV file name',
        metavar = 'file'
    ),
    make_option(
        c('-o', '--output'),
        type = 'character',
        default = NULL, 
        help = 'Name of the output file',
        metavar = 'output'
    ),
    make_option(
        c('-x', '--identifier'),
        type = 'character',
        default = TRUE, 
        help = 'Set the identifier of the original file captured by Darshan DXT',
        metavar = 'identifier'
    )
)
 
opt_parser = OptionParser(option_list=option_list)
opt = parse_args(opt_parser)

base = commandArgs()
base = base[grepl('--file', base, fixed=TRUE)]
if (!is.null(base)) {
    base = strsplit(base, split='=')
    base = sapply(base, tail, 1)
}

base = dirname(base)

df <- read.csv(file=opt$file, sep = ',')

df$duration = df$end - df$start

if (nrow(df) == 0) {
    quit()
}

df$label = paste0(
    'Rank: ', df$rank, '\n',
    'Operation: ', df$operation, '\n',
    'Duration: ', round(df$duration, digits = 3), ' seconds\n',
    'Size: ', (df$size / 1024), ' KB\n',
    'Offset: ', df$offset, '\n',
    'Lustre OST: ', ifelse(is.na(df$ost), '-', df$ost)
)

# Include a zero record to ensure we can facet the plot
df <- rbind(df, 
    data.frame(
        file_id = c(0, 0),
        api = c('POSIX', 'POSIX'),
        rank = c(0, 0),
        operation = c('write', 'read'),
        segment = c(0, 0),
        offset = c(0, 0),
        size = c(0, 0),
        start = c(0, 0),
        end = c(0, 0),
        duration = c(0, 0),
        label = c('', ''),
        ost = c(NA, NA)
    )
)

df$operation <- as.factor(df$operation)

palette <- wes_palette('Zissou1', 100, type = 'continuous')

maximum = max(df$offset) + (max(df$offset) * 0.01)

plot_posix_write <- ggplot(
    df[df$api == 'POSIX' & df$operation == 'write', ],
    aes(
        xmin = offset,
        xmax = offset + size,
        y = rank,
        color = size,
        text = label
    )) +
    geom_errorbarh(height=0) +
    #geom_segment() +
    scale_x_continuous(breaks = seq(0, maximum, length.out = 10)) +
    facet_grid(operation ~ .) +
    scale_color_gradientn(
        'Request size\n(bytes)',
        colours = palette
    ) +  
    expand_limits(x = 0) +
    ylim(0, max(df$rank)) +
    xlab('File offset (bytes)') +
    ylab('Rank #') +
    theme_bw() +
    theme(
        legend.position = "top",
        plot.title = element_text(size = 10),
        strip.background = element_rect(colour = NA, fill = NA)
    )

plot_posix_read <- ggplot(
    df[df$api == 'POSIX' & df$operation == 'read', ],
    aes(
        xmin = offset,
        xmax = offset + size,
        y = rank,
        color = size,
        text = label
    )) +
    geom_errorbarh(height=0) +
    #geom_segment() +
    scale_x_continuous(breaks = seq(0, maximum, length.out = 10)) +
    facet_grid(operation ~ .) +
    scale_color_gradientn(
        'Request size\n(bytes)',
        colours = palette
    ) +  
    expand_limits(x = 0) +
    ylim(0, max(df$rank)) +
    xlab('File offset (bytes)') +
    ylab('Rank #') +
    theme_bw() +
    theme(
        legend.position = "top",
        plot.title = element_text(size = 10),
        strip.background = element_rect(colour = NA, fill = NA)
    )

p_posix_write <- ggplotly(
        plot_posix_write,
        width = 1800,
        height = 1000,
        tooltip = "text",
        legendgroup = operation,
        dynamicTicks = TRUE
    ) %>%
    layout(
        margin = list(pad = 0),
        legend = list(orientation = "h", x = 0, y = length(df$ranks) + 6),
        autosize = TRUE,
        xaxis = list(title = 'File offset (bytes)', matches = 'x'),
        yaxis = list(title = 'Rank', matches = 'y', fixedrange = FALSE),
        hoverlabel = list(font = list(color = 'white')),
        title = paste0(
            'Explore <b>Spatiality</b>',
            '<br>',
            '<sup>',
            opt$identifier,
            '</sup>'
        )
    ) %>%
    style(
        showlegend = FALSE
    ) %>%
    toWebGL()

p_posix_read <- ggplotly(
        plot_posix_read,
        width = 1800,
        height = 1000,
        tooltip = "text",
        legendgroup = operation,
        dynamicTicks = TRUE
    ) %>%
    rangeslider(0, maximum, thickness = 0.03) %>%
    layout(
        margin = list(pad = 0),
        legend = list(orientation = "h", x = 0, y = length(df$ranks) + 6),
        autosize = TRUE,
        xaxis = list(title = 'File offset (bytes)', matches = 'x'),
        yaxis = list(title = 'Rank', matches = 'y', fixedrange = FALSE),
        hoverlabel = list(font = list(color = 'white'))
    ) %>%
    toWebGL()

p <- subplot(
    p_posix_write, p_posix_read,
    nrows = 2,
    titleY = TRUE,
    titleX = TRUE,
    shareX = TRUE,
    shareY = TRUE
) %>%
layout (
    margin = list(t = 130),
    images = list(
        source = raster2uri(as.raster(readPNG(paste(base, 'dxt-explorer.png', sep = '/')))),
        x = 0,
        y = 1.02, 
        sizex = 0.2,
        sizey = 0.2,
        xref = 'paper',
        yref = 'paper', 
        xanchor = 'middle',
        yanchor = 'bottom'
    )
) %>%
config(
    displaylogo = FALSE
)

saveWidget(p, selfcontained = self_contained, opt$output)

File: explorer/plots/spatiality.py
import os
import explorer
import numpy as np
import plotly.express as px
import pyarrow.feather as feather

from PIL import Image
from optparse import OptionParser


parser = OptionParser()
parser.add_option(
    "-f",
    "--file",
    type="string",
    default=None,
    help="DXT CSV file name",
    metavar="FILE",
)
parser.add_option(
    "-o",
    "--output",
    type="string",
    default=None,
    help="Name of the output file",
    metavar="output",
)
parser.add_option(
    "-x",
    "--identifier",
    type="string",
    default=None,
    help="Set the identifier of the original file captured by Darshan DXT",
    metavar="identifier",
)

(options, args) = parser.parse_args()
options = vars(options)

df = feather.read_feather(options["file"])

if df.empty:
    quit()

df["duration"] = df["end"] - df["start"]
rank_gap = max(df["rank"]) * 0.075
maximum_rank = max(df["rank"])

if len(df.index) == 0:
    quit()


def paste0():
    label = (
        "Rank: "
        + df["rank"].apply(str)
        + "<br>"
        + "Operation: "
        + df["operation"].apply(str)
        + "<br>"
        + "Duration: "
        + round(df["duration"], 3).apply(str)
        + " seconds<br>"
    )
    label = (
        label
        + "Size: "
        + (df["size"] / 1024).apply(str)
        + " KB<br>"
        + "Offset: "
        + df["offset"].apply(str)
        + "<br>"
        + "Lustre OST: "
    )

    if df["osts"] is None:
        label = label + "-"
    else:
        label = label + df["osts"].apply(str)
    return label


df["label"] = paste0()

df = df[df["api"] == "POSIX"]

conditions = [
    (df["size"] >= 0) & (df["size"] <= 100),
    (df["size"] >= 101) & (df["size"] <= 1000),
    (df["size"] >= 1001) & (df["size"] <= 10000),
    (df["size"] >= 10001) & (df["size"] <= 100000),
    (df["size"] >= 100001) & (df["size"] <= 1000000),
    (df["size"] >= 1000001) & (df["size"] <= 4000000),
    (df["size"] >= 4000001) & (df["size"] <= 10000000),
    (df["size"] >= 10000001) & (df["size"] <= 100000000),
    (df["size"] >= 100000001) & (df["size"] <= 10000000000),
    (df["size"] >= 10000000001),
]

values = ["0", "1", "2", "3", "4", "5", "6", "7", "8", "9"]

df["bin"] = np.select(conditions, values)

fig = px.scatter(
    df,
    x="offset",
    y="rank",
    color="bin",
    range_y=(0 - rank_gap, maximum_rank + rank_gap),
    error_x="size",
    facet_row="operation",
    custom_data=["label"],
    template="plotly_white",
    category_orders={"bin": ["0", "1", "2", "3", "4", "5", "6", "7", "8", "9"]},
)

col_names = {
    "0": "0-100",
    "1": "101-1K",
    "2": "1K-10K",
    "3": "10K-100K",
    "4": "100K-1M",
    "5": "1M-4M",
    "6": "4M-10M",
    "7": "10M-100M",
    "8": "100M-1G",
    "9": "1G+",
}
fig.for_each_trace(lambda t: t.update(name=col_names[t.name]))

fig.update_xaxes(showline=True, linewidth=1, linecolor="black", mirror=True)
fig.update_yaxes(showline=True, linewidth=1, linecolor="black", mirror=True)

fig.update_traces(
    error_x=dict(
        width=0,
    ),
    marker=dict(size=1),
    error_x_symmetric=False,
    hovertemplate="<br>".join(
        [
            "%{customdata[0]}",
        ]
    ),
)


path = os.path.abspath(explorer.__file__)
path = path.split("__init__.py")[0]
pyLogo = Image.open(path + "plots/dxt-explorer.png")

fig.add_layout_image(
    dict(
        source=pyLogo,
        xref="paper",
        yref="paper",
        x=0,
        y=1.15,
        sizex=0.2,
        sizey=0.2,
        xanchor="left",
        yanchor="top",
    )
)

fig.for_each_yaxis(lambda yaxis: yaxis.update(title="Rank"))

for annotation in fig.layout.annotations:
    if "write" in annotation.text:
        annotation.text = "Write"
    elif "read" in annotation.text:
        annotation.text = "Read"

fig.update_layout(
    legend=dict(
        itemsizing="constant",
        orientation="h",
        yanchor="bottom",
        y=1.008,
        xanchor="right",
        x=0.98,
    ),
    legend_title="Request Size",
    autosize=False,
    height=1200,
    width=1800,
    margin=dict(r=20, l=20, b=75, t=125),
    title=("Explore <b>Spatiality</b> <br>" + options["identifier"]),
    title_x=0.5,
    title_y=0.98,
    font=dict(size=13, color="#000000"),
    xaxis_title="Runtime (Seconds)",
    xaxis=dict(
        rangeslider=dict(visible=True),
        type="-",
    ),
    xaxis_rangeslider_thickness=0.04,
)

fig.write_html(options["output"])
